glib-macros/tests/test.rs

```diff
@@ -343,36 +343,29 @@ fn derive_variant() {
 #[test]
 fn closure() {
     let empty = glib::closure!(|| {});
-    assert!(empty.invoke(&[]).is_none());
+    empty.invoke::<()>(&[]);
 
     let no_arg = glib::closure!(|| 2i32);
-    assert_eq!(no_arg.invoke(&[]).unwrap().get::<i32>().unwrap(), 2);
+    assert_eq!(no_arg.invoke::<i32>(&[]), 2);
 
     let add_1 = glib::closure!(|x: i32| x + 1);
-    assert_eq!(add_1.invoke(&[&3i32]).unwrap().get::<i32>().unwrap(), 4);
+    assert_eq!(add_1.invoke::<i32>(&[&3i32]), 4);
 
     let concat_str = glib::closure!(|s: &str| s.to_owned() + " World");
-    assert_eq!(
-        concat_str
-            .invoke(&[&"Hello"])
-            .unwrap()
-            .get::<String>()
-            .unwrap(),
-        "Hello World"
-    );
+    assert_eq!(concat_str.invoke::<String>(&[&"Hello"]), "Hello World");
 
     let weak_test = {
         let obj = glib::Object::new::<glib::Object>(&[]).unwrap();
 
         assert_eq!(obj.ref_count(), 1);
         let weak_test = glib::closure_local!(@watch obj => move || obj.ref_count());
         assert_eq!(obj.ref_count(), 1);
-        assert_eq!(weak_test.invoke(&[]).unwrap().get::<u32>().unwrap(), 3);
+        assert_eq!(weak_test.invoke::<u32>(&[]), 3);
         assert_eq!(obj.ref_count(), 1);
 
         weak_test
     };
-    assert!(weak_test.invoke(&[]).is_none());
```

Comment:
Now this one is interesting. Invalidated closures return nothing too, even if the closure before had an actual return type that was not `()`.

I think we'll need to have `invoke()` return a `Result` and check for invalidation afterwards to distinguish the two no-result cases.

Reply:
Actually worse: the return value has to be initialized to the correct type before invoking and we currently don't do that. And then it's impossible to distinguish `()` from any other type as passing another type in will simply keep it at that type.

This only works for closures we create right now because we also initialize the type of the value when returning.

Also for closures from C the arguments are apparently not type-checked at runtime either.

I think `invoke()` has to be an `unsafe` function, unfortunately.

----

For signal closures this is no problem because there the return type and the arguments are introspectable at runtime. However when connecting a `Closure` as signal handler there will only be criticals at runtime if the return type does not match.

FYI @jf2048

Reply:
The main problem  I see here now is that `connect_closure()` is actually unsafe. If your closure returns a `String` and the C code assumes an `Object`, then it would silently use the string pointer as an object pointer and things fall apart.

That's why we have the additional checks inside the `GClosure` closures inside the normal `connect()`.

Reply:
I.e. the whole `closure!` and `connect_closure()` machinery is currently not good

Reply:
I think we'd go with having a `struct RustClosure { closure: glib::Closure, return_type: glib::Type, parameters: Vec<glib::Type> }` from the macro (the constructor of that would have to be unsafe) and then make use of that in `connect_closure()` to check if things match up.

`RustClosure` would have a bit of safe API (including `invoke()` but not the constructor for obvious reasons), and all `glib::Closure` related API would be unsafe

Reply:
I have this half-implemented now, just need to put all the pieces together.

Reply:
Hrm, I have some problems coming up with a way that infers the closure argument/return types while also working for all the different special `@` captures we have. I have one version that compiles and causes a memory leak for `@strong`, and versions that fail compiling on one or another for the closure variants.